core/field.ts

```diff
@@ -1046,8 +1056,7 @@ export abstract class Field<T = unknown> implements IASTNodeLocationSvg,
    * @param opt_newValue The value to be validated.
    * @returns The validated value, same as input by default.
    */
-  protected doClassValidation_(opt_newValue?: AnyDuringMigration):
-      AnyDuringMigration {
+  protected doClassValidation_(opt_newValue?: T|null): T|null {
```

Comment:
If it should support any value but we don't know what the value is, instead of `any` we should use `unknown` to enforce properly handling parsing the value. I'm not sure if this is the best option, though. The current implementation doesn't support actually being able to properly parse an unknown value to some T value at the root since the root doesn't know what T is. We would need to require all subclasses to provide implementation for parsing `T` from `unknown`.

I also don't think `unknown` is actually desirable. One major benefit of adding type information is it simplifies using the class by narrowing the information space. For example, a `FieldNumber` that accepts `string|number` tells the implementer at compile time that this will probably be valid if they supply those types. If it accepts `unknown`, the implementer might supply a `boolean` and won't know it's an issue until debugging it at runtime.

Reply:
I brought this up in the team meeting on Friday, and @rachel-fenichel reminded me that we could overload the function properties! We can use `T` to guide the caller of `doClassValidation_` in providing the right value at compile time, though we require `unknown` to guide the implementer of `doClassValidation_` to properly handle all cases.

```ts
/**
   * Used to validate a value. Returns input by default. Can be overridden by
   * subclasses, see FieldDropdown.
   *
   * @param opt_newValue The value to be validated.
   * @returns The validated value, same as input by default.
   */
  protected doClassValidation_(opt_newValue?: T|null): T|null;
  protected doClassValidation_(opt_newValue?: unknown): unknown|null {
    if (opt_newValue === null || opt_newValue === undefined) {
      return null;
    }
    // NOTE: `unknown` cannot be parsed to T here since we don't know what
    // T is, though subclasses should handle validating any type of input.
    return opt_newValue as T;
  }
  ```
  
  What do you think?

Reply:
Sorry, GitHub keeps hiding this conversation and I missed this reply.

So I have two main problems still.

First, is about why we have the class validator to begin with. At compile time, we don't necessarily know that the value will be type T, so this could still be problematic. If we did know it would be type T, then we wouldn't really need this validator, as the point of it is to make sure the value is the correct type. As an extreme example, Neil was working on a block a while back that let you `eval` arbitrary code. We don't know what the type of that value would be.

Second, we're making assumptions about the types of both the user input and the value, that I don't think we can make. Here's a concrete example: https://github.com/google/blockly-samples/blob/master/examples/pitch-field-demo/field_pitch.js#LL236C29-L236C29
The field's value is always a number, but what we're taking as a parameter here is a string. So the type of the field's value doesn't necessarily match the type of the parameter in the validator.

Or let's say you wanted to subclass FieldNumber, but instead of the user typing in a number, they upload an image, and in the class validator we look at the image and use OCR to convert the image to a number. If we narrow the types right here, would you be able to do this in a subclass of FieldNumber? I don't think you would, but if I'm wrong and it works, then I do have fewer concerns about this approach, but not none. If fields were designed differently, where the user input were separate from the actual data in the field, this wouldn't be a big deal. But they're not, and historically we've let `setValue` and the others take in arbitrary values, so imo we need to keep allowing that.

Reply:
As discussed on the call earlier today, we landed on providing a `U` type that supports cases like this where the input of `setValue` may not be compatible with `T`, the value being stored.